### PR TITLE
Rewrite filter push down

### PR DESCRIPTION
**Pull request by humengyu2012**
Do not merge this PR, because there are some documents to write.

### CODE REVIEW

**Comment by humengyu2012** [comment]
> Do not merge this PR, because there are some documents to write.

Docs was added.

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

File: flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java [hunk 1]
```diff
@@ -60,151 +69,321 @@
...
+          return Optional.empty();
       }
     }
-    if (resolvedExpression instanceof FieldReferenceExpression) {
-      String name = ((FieldReferenceExpression) resolvedExpression).getName();
-      return Expressions.column(name, nameTypeMap.get(name));
+    return Optional.empty();
+  }
+
+  // We use empty as placeholder, which may cause the range of filters to become larger.
```
[the remark this answers is not in the document]
Comment (humengyu2012) [reply flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java:137]: Placeholder can be anything, and empty can help us deal with it without forgetting.

File: flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java [hunk 2]
```diff
@@ -60,151 +69,321 @@
...
+
+    final String name = columnElement.getColumnName();
+    final DataType type = columnElement.getType();
+    final DataType castType = columnElement.getCastType();
+    final org.tikv.common.types.DataType tidbType = nameTypeMap.get(name);
+    final Object value = valueElement.getValue();
+    final DataType valueType = valueElement.getType();
+
+    // We will replace 'CAST(`c1` as int) = 1' with '`c1` = 1',
+    // which means that 'CAST' syntax will be ignored, please use '`c1` = CAST(1 AS INT)'.
```
[the remark this answers is not in the document]
Comment (humengyu2012) [reply flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java:226]: There is no expression of `cast` for tikv. If we want support expression like `'CAST(`c1` as int) = 1'`, we must cast 1 as        `c1.datatype`, it is not a good idea which may cause errors.

File: flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java [hunk 3]
```diff
@@ -60,151 +69,321 @@
...
+    ColumnElement columnElement =
+        leftElement instanceof ColumnElement ? (ColumnElement) leftElement
+            : (ColumnElement) rightElement;
+
+    final String name = columnElement.getColumnName();
+    final DataType type = columnElement.getType();
+    final DataType castType = columnElement.getCastType();
+    final org.tikv.common.types.DataType tidbType = nameTypeMap.get(name);
+    final Object value = valueElement.getValue();
+    final DataType valueType = valueElement.getType();
```
[the remark this answers is not in the document]
Comment (humengyu2012) [reply flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java:223]: Reserved for new feature, such as set, enum and json. But these types are not supported now by tikv.

File: flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java [hunk 4]
```diff
@@ -60,151 +69,321 @@
...
+      DataTypes.TINYINT(),
+      DataTypes.SMALLINT(),
+      DataTypes.INT(),
+      DataTypes.BIGINT()
+  );
 
-  FilterPushDownHelper(ResolvedCatalogTable table) {
-    this.table = table;
-  }
+  private final TiTableInfo tiTableInfo;
```
[the remark this answers is not in the document]
Comment (humengyu2012) [reply flink/flink-1.13/src/main/java/io/tidb/bigdata/flink/connector/table/FilterPushDownHelper.java:80]: I am not sure whether we will need it in the future, so i decide to reserve it.

**Review by xuanyu66 — APPROVED** [review]

**Review by sunxiaoguang — APPROVED** [review]

**Pull request merged**